lockfile/src/cyclonedx.rs

```diff
@@ -164,23 +164,37 @@ fn from_purl<T: Component>(component: &T) -> anyhow::Result<Package> {
 
 pub struct CycloneDX;
 
+impl CycloneDX {
+    fn process_components<T: Component>(components: Option<&[T]>) -> anyhow::Result<Vec<Package>> {
+        components.map_or(Ok(Vec::new()), |comp| {
+            let component_iter = filter_components(comp);
+            let mut packages = Vec::new();
```

Comment:
I don't know if it's better or not, but this can be rewritten functionally.

```rust
let packages = filter_components(comp)
    .filter_map(|comp| match from_purl(comp) {
        Ok(pkg) => Some(Ok(pkg)),
        Err(e) if e.is::<UnknownEcosystem>() => {
            log::warn!("{:?}", e);
            None
        },
        Err(e) => Some(Err(e)),
    })
    .collect::<Result<Vec<_>>>()?;
```

Or maybe this way, but I wish there was an `is_ok_or`:

```rust
let packages = filter_components(comp)
    .map(from_purl)
    .filter(|r| !r.is_err_and(|e| e.is::<UnknownEcosystem>()))
    .collect::<Result<Vec<_>>>()?;
```

Without the logging, it's very simple.